### DISCUSSION TITLE
Using index_unit while creating new LAS

### DISCUSSION THREAD

**Discussion by geoms4geos** [kinverarity1/lasio#538] [Q&A]
I've been a long time user of lasio and love the functionality it adds to python for those of us who utilize las files for import into various software packages. With that said I continue to struggle with one issue that I believe must have a simple fix. How do I initialize a blank las file with "ft" as the depth index units? I have always assumed the following code would give me correct units:

`las.LASFile(index_unit = 'ft') `

What am I missing? I have fooled lasio into doing my bidding by changing the units in defaults to 'ft', but that is just a cheap hack. How do I do this the correct way?

**Comment by dcslagel** [comment]
@geoms4geos ,

I rummaged through the test files for an answer. It looks like this code converts the depth unit from meters to  depth unit in feet.

`las = lasio.read("autodepthindex_M.las"), index_unit="f")`


Notice the index_unit is "f" instead of "ft".  So I think you could try `las.LASFile(index_unit = 'f')` and see if that does the job.

**Reply by geoms4geos** [reply]
@dcslagel Thanks for digging. We have been using the index_unit argument to read existing las files and to load curves into las files. Our experience is that it has been kind of hit and miss on working to flag the index/depth curve as feet. 

Setting the defaults to 'ft' has been much more consistent for making new las files with lasio.LASFile(). Having to hardcode defaults like that makes me feel like I am missing a kwarg somewhere, but I am stumped.

**Reply by geoms4geos** [reply]
@dcslagel Following up a bit more, this code I played with last night is what I find most interesting. 

```
import lasio

las = lasio.LASFile(index_unit='ft')
print('\nLAS INDEX UNIT = {0}'.format(las.index_unit))
#print(las.header)

las.index_unit = 'ft'
print('\nLAS INDEX UNIT = {0}'.format(las.index_unit))
#print(las.header)
```

When we write a brand new file with .LASFile() it appears to me that no index_unit is created. We can define index_unit later and it is populated. This is what makes me think I am missing a kwarg when I generate the new las file.

**Reply by dcslagel** [reply]
Looking more fully.... I'll have an update in a day or so.

**Comment by dcslagel** [comment]
@geoms4geos ,

Okay, I looked through the code, code-history, and documentation.  The way that index_unit has evolved so far, it doesn't yet have the functionality to set the index_unit when initializing a LAS object for building a LAS file from scratch.   

It does seem like functionality that should be added. If you agree, one of us can click the "Create issue from discussion" link on this page and that will create an issue to track the change-request.

In the current version of Lasio (0.30), what you are doing is probably the best way to quickly enable the creation of a LAS file with feet as the initial index_unit.  I'm guessing. you are changing the index_unit from "m" to "f" in the section of defaults.py that has the Section items.  Others can to that for now if they have the same need.

```python
       "Well": SectionItems(                                                                                                                                                                                                                                                                                      
              [                                                                                                                                                                                                                                                                                                      
                 HeaderItem("STRT", "m", np.nan, "START DEPTH"),                                                                                                                                                                                                                                                    
                 HeaderItem("STOP", "m", np.nan, "STOP DEPTH"),                                                                                                                                                                                                                                                     
                 HeaderItem("STEP", "m", np.nan, "STEP"),        
                ...
             ]
     )
```  

--------------------------------------

#### Notes for enabling setting the index_unit when initializing a LAS object for building a LAS file from scratch:
- I think the initial code location is las.py::init()
- After the default template is read in then if the index_unit keyword parameter has a valid value then the LAS object needs to be updated with the index_unit value.
- The documentation should be updated to explain the index_unit behavior:  `docs/source/building.rst`

**Comment by kinverarity1** [comment]
Thanks @dcslagel - spot on. I never anticipated this use case. I've created #542.

**Reply by geoms4geos** [reply]
@dcslagel Even more great digging! You are spot on, I have gone into my defaults.py file and changed the default units for depth references as below. 

```
"Well": SectionItems(
            [
                HeaderItem("STRT", "ft", np.nan, "START DEPTH"),
                HeaderItem("STOP", "ft", np.nan, "STOP DEPTH"),
                HeaderItem("STEP", "ft", np.nan, "STEP"),
                HeaderItem("NULL", "", -9999.25, "NULL VALUE"),
                HeaderItem("COMP", "", "", "COMPANY"),
                HeaderItem("WELL", "", "", "WELL"),
                HeaderItem("FLD", "", "", "FIELD"),
                HeaderItem("LOC", "", "", "LOCATION"),
                HeaderItem("PROV", "", "", "PROVINCE"),
                HeaderItem("CNTY", "", "", "COUNTY"),
                HeaderItem("STAT", "", "", "STATE"),
                HeaderItem("CTRY", "", "", "COUNTRY"),
                HeaderItem("SRVC", "", "", "SERVICE COMPANY"),
                HeaderItem("DATE", "", "", "DATE"),
                HeaderItem("UWI", "", "", "UNIQUE WELL ID"),
                HeaderItem("API", "", "", "API NUMBER"),
            ]
```
@kinverarity1 I also tested out the current functionality to understand when index_unit and HeaderItems get set. The following code seems to indicate that once an index_curve is created, HeaderItems can be set to "ft" using update_units_from_index_curve(), but even in this event, index_unit is not inherently set for the las object: 

```
import lasio

las2 = lasio.LASFile()
depths = np.arange(10, 50, 0.5)
las2.add_curve('DEPT', depths, unit = 'ft')
las2.update_units_from_index_curve()
```

I'm playing with some code to introduce an index_unit kwarg into LASFile() but I still have some debugging to do.
